Delete storage feature

Nagendracv opened the pull request:
Delete storage feature (Draft)

## Description
Added ability to delete a storage blob. The mechanism used is to delete by deleting the tf file and committing the changes.

## How Has This Been Tested?
Yes, tested until the delete succeeds in azure. The Output handler queue has not been tested.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Unit test (new or update to tests)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (change that would cause documentation to be updated)
- [ ] Configuration change (change that would affect the configuration of the application)

## Checklist:
- [x] PR is submitted to the correct branch `develop`.
- [x] Code follows the code style of this project.
- [ ] Changes are covered by Specflow tests and all new or existing tests are passing.
- [ ] Any new or updated translatable strings have been added to the localization files.
- [x] Necessary and relevant documentation has been created or updated.
- [x] There are either no changes to the application's configuration or the necessary changes in the infrastructure repository are included in a separate PR.

Nagendracv committed:
Delete storage support (Draft)

Nagendracv committed:
logic update to delete all files related to storage instead of just the tf file

yjmrobert commented on Portal/src/Datahub.Functions/TerraformOutputHandler.cs:
```diff
@@ -393,4 +402,39 @@ private static string GetStatusMapping(string value)
+        }
+
+        return true;
+    }
```
Let's do a soft delete like we discussed, where the row isn't removed but the state is marked as deleted

Nagendracv replied:
soft delete implemented now

yjmrobert commented on Portal/src/Datahub.Infrastructure/Services/RequestManagementService.cs:
```diff
@@ -125,6 +126,13 @@ await ctx.Entry(project)
                 await _resourceMessagingService.GetWorkspaceDefinition(project.Project_Acronym_CD,
                     requestingUser.Email);
+
+            if (isDelete) {
+                foreach (var template in allTemplates) {
+                    if (template.Name == terraformTemplate) template.Status = "delete";
+                }
+            }
+            
```
The status here won't be persisted to the database. It would make more sense to process this up near L108 so that the `ProcessRequest` call can persist the template status into the database

Nagendracv replied:
now persisted

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor:
```diff
@@ -64,7 +63,11 @@
                                     <MudButton Variant="@Variant.Filled" Color="@Color.Primary" Disabled>
                                         @Localizer["Tool has been created"]
-                                        <MudIcon Icon="fa-light fa-check-to-slot" Class="ml-2" Style="font-size: 0.8rem;"/>
+                                        <MudIcon Icon="fa-light fa-check-to-slot" Class="ml-1" Style="font-size: 0.8rem;"/>
```
I think a delete style icon might be more appropriate. We have the font-awesome pro library so there are a lot of options

Nagendracv replied:
used Trash Icon

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor:
```diff
@@ -120,7 +123,9 @@
     {
         Available,
         SendingRequest,
+        SendingDelete,
         PendingRequest,
+        PendingDeleteRequest,
```
Are there UI updates for these statuses? I can see the updates but no corresponding blazor

Nagendracv replied:
Updated razor code as well

yjmrobert commented on ResourceProvisioner/src/ResourceProvisioner.Infrastructure/Services/TerraformService.cs:
```diff
@@ -327,4 +361,43 @@ private string ComputeBackendConfigValue(string workspaceName, string variableNa
     }
+
+    public async Task DestroySpecificResourcesAsync(TerraformWorkspace terraformWorkspace, string resourceIdentifier)
+    {
+        var projectPath = DirectoryUtils.GetProjectPath(_resourceProvisionerConfiguration, terraformWorkspace.Acronym);
+
+        Directory.SetCurrentDirectory(projectPath);
+
+        if (!await ExecuteShellCommand("terraform init")) {
+            _logger.LogError("Terraform init failed.");
+            return;
+        }
+
+        var destroyCmd = $"terraform destroy -auto-approve -target={resourceIdentifier}";
+        if (!await ExecuteShellCommand(destroyCmd)) {
+            _logger.LogError($"Terraform destroy failed for {resourceIdentifier}.");
+        }
+
+        _logger.LogInformation("Successfully destroyed specified resources.");
+    }
+    private async Task<bool> ExecuteShellCommand(string command)
+    {
+        using var process = new System.Diagnostics.Process
+        {
+            StartInfo = new System.Diagnostics.ProcessStartInfo
+            {
+                FileName = "cmd.exe",
+                Arguments = $"/c {command}",
+                RedirectStandardOutput = true,
+                UseShellExecute = false,
+                CreateNoWindow = true,
+            }
+        };
+
+        process.Start();
+        await process.WaitForExitAsync();
+
+        return process.ExitCode == 0;
+    }
+
```
Comment out

Nagendracv replied:
commented now

Nagendracv committed:
Fixed test error

Nagendracv committed:
one missing test location

Nagendracv committed:
Merged

Nagendracv committed:
Updated check for deleted in all locations where resource is queried

Nagendracv committed:
Some more queries updated

Nagendracv committed:
build error due to space

Nagendracv committed:
using enum

Nagendracv committed:
Solution file update

Nagendracv committed:
Commented mehtod for later use

Nagendracv committed:
icon updated

Nagendracv committed:
Delete Icon / Button updates

Nagendracv committed:
Spinner

Nagendracv committed:
Merged with Develop

Nagendracv committed:
delete unit test

Nagendracv committed:
Merged

yjmrobert commented:
@Nagendracv, just to confirm before I review. This feature is complete and were I to run it I would be able to provision a storage for my workspace and delete the storage?

Nagendracv committed:
Localized status names, semaphore fix and dev appsettings fix

Nagendracv committed:
Merged from Develop

Nagendracv committed:
Merge corrected

Nagendracv commented:
> @Nagendracv, just to confirm before I review. This feature is complete and were I to run it I would be able to provision a storage for my workspace and delete the storage?

Yes full cycle is now tested - including the datahub.functions - terraform output handler.
Earlier I had tested with local data using the sudo page expected output. 
Now connected with the azure deployed dev application's output storage queue.
No changes were needed there as well

Nagendracv committed:
Made Delete button look nicer. Also fixed french

yjmrobert commented on Datahub.sln:
```diff
@@ -127,9 +130,19 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Datahub.Infrastructure.Offl
+		ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\requirements.txt = ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\requirements.txt
+		ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\workspace_data.json = ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\workspace_data.json
+	EndProjectSection
 EndProject
```
Not part of this PR. I would recommend using vscode when working with python. No need to add the files to the csharp solution items

Nagendracv replied:
Done. I had added it here earlier for some reason. But sure it is not needed in the solution.

yjmrobert commented on Portal/src/Datahub.Core/Utils/TerraformOutputHelper.cs:
Please refactor all `(r.Status == null || r.Status != TerraformOutputStatus.Deleted))` into a property method

yjmrobert commented on Portal/src/Datahub.Functions/TerraformOutputHandler.cs:
```diff
@@ -236,17 +236,24 @@ internal async Task ProcessAzureStorageBlob(IReadOnlyDictionary<string, Terrafor
         var storageBlobStatus =
             GetStatusMapping(outputVariables[TerraformVariables.OutputAzureStorageBlobStatus].Value);
-        if (!storageBlobStatus.Equals(TerraformOutputStatus.Completed, StringComparison.InvariantCultureIgnoreCase))
-        {
-            _logger.LogError("Azure storage blob status is not completed. Status: {Status}", storageBlobStatus);
+        if (!storageBlobStatus.Equals(TerraformOutputStatus.Completed, StringComparison.InvariantCultureIgnoreCase)) {
+            if (!storageBlobStatus.Equals(TerraformOutputStatus.Deleted, StringComparison.InvariantCultureIgnoreCase)) {
+                _logger.LogError("Azure storage blob status is failed. Status: {Status}", storageBlobStatus);
+                return;
+            }
+            bool isDeleted = await DeleteProjectResource(outputVariables, storageServiceType);
+            if (isDeleted)
+                _logger.LogInformation("Azure storage blob status is Deleted. Status: {Status}", storageBlobStatus);
+            else
+                _logger.LogError("Azure storage blob deleted from azure but could not be removed from project. Status: {Status}", storageBlobStatus);
             return;
         }
 
-        var projectResource = await GetProjectResource(outputVariables,
-            TerraformTemplate.GetTerraformServiceType(TerraformTemplate.AzureStorageBlob));
+        var projectResource = await GetProjectResource(outputVariables, storageServiceType);
```
Please simplify the nesting

yjmrobert commented on Portal/src/Datahub.Functions/TerraformOutputHandler.cs:
```diff
@@ -391,6 +400,44 @@ private static string GetStatusMapping(string value)
+
         return projectResource;
     }
+    private async Task<bool> DeleteProjectResource(IReadOnlyDictionary<string, TerraformOutputVariable> outputVariables, string terraformServiceType)
+    {
+        try {
+            
+            var projectAcronym = outputVariables[TerraformVariables.OutputProjectAcronym];
+        
+            var project = await _projectDbContext.Projects.Include(p => p.Resources)
+                                              .FirstOrDefaultAsync(p => p.Project_Acronym_CD == projectAcronym.Value);
+            if (project is null) {
+                _logger.LogError("Project not found for acronym {ProjectId}", projectAcronym.Value);
+                throw new Exception($"Project not found for acronym {projectAcronym.Value}");
+            }
+
+            var projectResource = project.Resources
+                .FirstOrDefault(x => x.ResourceType == terraformServiceType);
+
+            if (projectResource is null) {
+                _logger.LogError("Project resource not found for project acronym {ProjectAcronymValue} and service type {TerraformServiceType}",
+                                  projectAcronym.Value, terraformServiceType);
+                throw new Exception($"Project resource not found for project acronym {projectAcronym.Value} and service type {terraformServiceType}");
+            }
+
+            projectResource.Status = TerraformOutputStatus.Deleted;
+            _projectDbContext.Project_Resources2.Update(projectResource);
+
+            await _projectDbContext.SaveChangesAsync();
+
+            _logger.LogInformation("Project resource with service type {TerraformServiceType} deleted successfully for project acronym {ProjectAcronymValue}",
+                                    terraformServiceType, projectAcronym.Value);
+            return true;
+        }
+        catch (Exception e) {
+            _logger.LogError(e, "Error deleting project resource");
+        }
+
+        return true;
+    }
```
In our previous call I suggested just putting L428 `projectResource.Status = storageBlobStatus` up on L277 instead of doing all this a second time inside a method, and fixing all the complex nesting.

Why was this not implemented?

Nagendracv replied:
Apologies, for some reason, this particular comment was not visible in my review list. I've updated it, will push in a minute

yjmrobert replied:
That's not what I said, your update never sets the status unless it is deleted.

![image](https://github.com/ssc-sp/datahub-portal/assets/43834308/bbcc403a-94ce-42d1-8341-f8ea2c5ba319)

it should be something like

![image](https://github.com/ssc-sp/datahub-portal/assets/43834308/5ce12ab5-925c-4fec-8721-96035ee9db3a)

Nagendracv replied:
No... why? Actually Status field would be used only in case of deleted... well mostly... unless you want to set the status field irrespective of deleted...

your code above is also correct - except that status field will always get set unless its either completed or deleted...
hmm... that amounts to the same thing too... well...

If we are going that route anyways, my thought is, probably, we can make it even more simple
i.e. always set the status field and let the output value from the terraform feed the status field

i.e. this would not be needed... coz it does not mean it has failed. It just means that the output status was set to some other value other than completed and deleted. So I believe we should not 'return' at this juncture, and let just set the status value be set. I think the value will be useful for later troubleshooting as well
![image](https://github.com/ssc-sp/datahub-portal/assets/19173124/4962ad32-d627-42e5-8b34-3f08a3a51af0)

What are your thoughts on that?

Nagendracv replied:
like this 
![image](https://github.com/ssc-sp/datahub-portal/assets/19173124/29aa8f16-839d-4ff3-ad64-fbe5ba17dd34)

Nagendracv replied:
In that case, it will set all values - including the deleted value. And I guess that's what we also want - especially to handle the failed scenario -  assuming there is another status called Failed status being set in the output value

yjmrobert reviewed (changes requested):
I stopped reviewing after `TerraformOutputHandler.cs`. I've provided feedback on that flow multiple times and it hasn't been implemented. 

Please revisit the feedback given and update

Nagendracv committed:
Update for nesting

Nagendracv committed:
solution file updated to remove python files

Nagendracv committed:
Removed code check for explicit delete. We will store all output status - including delete - which makes it a soft delete.

Nagendracv committed:
Updated delete within the block and return.
The original code was deleting the content inside the function. thats the only difference.

Nagendracv commented:
Closing the PR as it was meant only for the task for delete resource. Will do new PR once the whole User Story is complete
